pkg/querier/vcs.go

```diff
@@ -0,0 +1,664 @@
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/PuerkitoBio/goquery"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/google/go-github/v58/github"
+	giturl "github.com/kubescape/go-git-url"
+	"github.com/kubescape/go-git-url/apis"
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
+	"golang.org/x/mod/semver"
+	"golang.org/x/oauth2"
+	o2endpoints "golang.org/x/oauth2/endpoints"
+
+	vcsv1 "github.com/grafana/pyroscope/api/gen/proto/go/vcs/v1"
+	vcsv1connect "github.com/grafana/pyroscope/api/gen/proto/go/vcs/v1/vcsv1connect"
+	"github.com/grafana/pyroscope/pkg/querier/golang"
+	"github.com/grafana/regexp"
+)
+
+var (
+	githubClientID                                    = os.Getenv("GITHUB_CLIENT_ID")
+	githubClientSecret                                = os.Getenv("GITHUB_CLIENT_SECRET")
+	_                  vcsv1connect.VCSServiceHandler = (*Querier)(nil)
+)
+
+// todo better package structure.
+// vcs_service_test.go
+// vcs_service.go
+// source/find.go
+// source/golang/find.go
+
+func (q *Querier) GithubApp(ctx context.Context, req *connect.Request[vcsv1.GithubAppRequest]) (*connect.Response[vcsv1.GithubAppResponse], error) {
+	return connect.NewResponse(&vcsv1.GithubAppResponse{
+		ClientID: githubClientID,
+	}), nil
+}
+
+func (q *Querier) GithubLogin(ctx context.Context, req *connect.Request[vcsv1.GithubLoginRequest]) (*connect.Response[vcsv1.GithubLoginResponse], error) {
+	auth, err := githubOauthConfig()
+	if err != nil {
+		return nil, err
+	}
+	token, err := auth.Exchange(ctx, req.Msg.AuthorizationCode)
+	if err != nil {
+		return nil, err
+	}
+	cookieValue, err := encryptAES256(token)
+	if err != nil {
+		return nil, err
+	}
+	// Sets a cookie with the encrypted token.
+	// Only the server can decrypt the cookie.
+	cookie := http.Cookie{
+		Name:  "GitSession",
+		Value: cookieValue,
+		// Refresh expiry is 6 months based on github docs
+		Expires:  time.Now().Add(15811200 * time.Second),
+		HttpOnly: true,
+		Secure:   true,
+		SameSite: http.SameSiteLaxMode,
+	}
+	resp := connect.NewResponse(&vcsv1.GithubLoginResponse{})
+	resp.Header().Add("Set-Cookie", cookie.String())
+	return resp, nil
+}
+
+func (q *Querier) GetFile(ctx context.Context, req *connect.Request[vcsv1.GetFileRequest]) (*connect.Response[vcsv1.GetFileResponse], error) {
+	gitURL, err := giturl.NewGitURL(req.Msg.RepositoryURL) // initialize and parse the URL
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	if gitURL.GetProvider() != apis.ProviderGitHub.String() {
+		return nil, connect.NewError(connect.CodeInvalidArgument, errors.New("only GitHub repositories are supported"))
+	}
+	cookie, err := (&http.Request{Header: req.Header()}).Cookie("GitSession")
+	if err != nil {
+		return nil, err
+	}
+	// todo: we can support multiple provider: bitbucket, gitlab, etc.
+	client, err := NewGithubClient(ctx, cookie)
```

Comment:
Why do we create a new client every time?

Reply:
Because the credentials are stored in an encrypted AES256 cookie and on every call we need to get that token. This is purely to make this simpler for now so that we don't need to store or cache tokens. In fact tokens are only owned by users.